torch/csrc/jit/codegen/cuda/kernel_ir.h

```diff
@@ -1169,15 +1250,55 @@ class TORCH_CUDA_CU_API TensorIndex final : public Val {
     return indices_;
   }
 
+  const std::vector<Val*>& strides() const {
+    return strides_;
+  }
+
   TensorView* view() const {
     TORCH_INTERNAL_ASSERT(view_ != nullptr);
     // TODO(kir): remove the need for const_cast
     return const_cast<fuser::cuda::kir::TensorView*>(view_); // NOLINT
   }
 
+  auto indexMode() const {
+    return index_mode_;
+  }
+
+  auto indexSwizzle() const {
+    return index_swizzle_;
+  }
+
+  auto& indexMode() {
+    return index_mode_;
+  }
+
+  auto& indexSwizzle() {
+    return index_swizzle_;
+  }
+
+  //! Indicates whether this index will be generated
+  //!  in 1D mode or 2D mode.
+  enum class TensorIndexMode { Linear1D, Linear2D };
+
+  //! All natively supported swizzling pattern are in tensor index
+  //!  for now and could build out and further formalize.
+  enum class TensorIndexSwizzle {
+    NoSwizzle = 0,
+
+    // mma swizzles
```

Comment:
Type system seems like it would be good here. It seems like you're specifying the "conversions" between layouts here, and if the smem TVs had "a type" we could specify those instead of the conversions in/out of the format.

Reply:
Yes this is currently specified in terms of conversion and I will think about the data layout formulation here. Maybe `Congruous` and `Crosswise` makes sense for gemm as a start but I will need to think a bit more on the layout categories we want to have.